How to access Microsoft Graph Java API from Java application behind a proxy server?

Expected behavior
Is there a way to provide proxy server configuration(proxy server url, username, password etc.) while accessing Microsoft Graph Java API from Java application behind a proxy server?
Actual behavior
We are using MS graph Java sdk in order to access O365 ODFB drives from our java application. When I am bypassing the enterprise proxy server(connected to open internet) then I am able to get desired response from the APIs. But when I am connecting via proxy server, I am getting the following exception:
SEVERE: Throwable detail: com.microsoft.graph.core.ClientException: Error during http request Exception in thread "main" com.microsoft.graph.core.ClientException: Error during http request at com.microsoft.graph.http.DefaultHttpProvider.sendRequestInternal(DefaultHttpProvider.java:356) at com.microsoft.graph.http.DefaultHttpProvider.send(DefaultHttpProvider.java:197) at com.microsoft.graph.http.DefaultHttpProvider.send(DefaultHttpProvider.java:177) at com.microsoft.graph.http.BaseCollectionRequest.send(BaseCollectionRequest.java:87) at com.microsoft.graph.requests.generated.BaseDriveItemCollectionRequest.get(BaseDriveItemCollectionRequest.java:54) at com.example.TestMain.main(TestMain.java:148) Caused by: java.net.ConnectException: Connection timed out: connect at java.net.DualStackPlainSocketImpl.connect0(Native Method) at java.net.DualStackPlainSocketImpl.socketConnect(Unknown Source) at java.net.AbstractPlainSocketImpl.doConnect(Unknown Source) at java.net.AbstractPlainSocketImpl.connectToAddress(Unknown Source) at java.net.AbstractPlainSocketImpl.connect(Unknown Source) at java.net.PlainSocketImpl.connect(Unknown Source) at java.net.SocksSocketImpl.connect(Unknown Source) at java.net.Socket.connect(Unknown Source) at sun.security.ssl.SSLSocketImpl.connect(Unknown Source) at sun.security.ssl.BaseSSLSocketImpl.connect(Unknown Source) at sun.net.NetworkClient.doConnect(Unknown Source) at sun.net.www.http.HttpClient.openServer(Unknown Source) at sun.net.www.http.HttpClient.openServer(Unknown Source) at sun.net.www.protocol.https.HttpsClient.<init>(Unknown Source) at sun.net.www.protocol.https.HttpsClient.New(Unknown Source) at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(Unknown Source) at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(Unknown Source) at sun.net.www.protocol.http.HttpURLConnection.plainConnect(Unknown Source) at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(Unknown Source) at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source) at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source) at java.net.HttpURLConnection.getResponseCode(Unknown Source) at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(Unknown Source) at com.microsoft.graph.http.UrlConnection.getResponseCode(UrlConnection.java:110) at com.microsoft.graph.http.DefaultHttpProvider.sendRequestInternal(DefaultHttpProvider.java:297)
I already asked a question with same title at stackoverflow:
https://stackoverflow.com/questions/53079388/how-to-access-microsoft-graph-java-api-from-java-application-behind-a-proxy-serv
Steps to reproduce the behavior
Try to access Graph Java SDK APIs behind a proxy server which requires authentication.
[AB#6049](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/6049)

--- a/src/main/java/com/microsoft/graph/authentication/IAuthenticationProvider.java
+++ b/src/main/java/com/microsoft/graph/authentication/IAuthenticationProvider.java
@@ -26,7 +26,9 @@ import com.microsoft.graph.http.IHttpRequest;
 
 /**
  * Provides authentication for a requests before it is sent by an HTTP provider
+ * @deprecated use ICoreAuthenticationProvider instead
  */
+@Deprecated
 public interface IAuthenticationProvider {
 
     /**

--- a/src/main/java/com/microsoft/graph/core/DefaultClientConfig.java
+++ b/src/main/java/com/microsoft/graph/core/DefaultClientConfig.java
@@ -32,6 +32,8 @@ import com.microsoft.graph.logger.ILogger;
 import com.microsoft.graph.serializer.DefaultSerializer;
 import com.microsoft.graph.serializer.ISerializer;
 
+import okhttp3.OkHttpClient;
+
 /**
  * The default configuration for a service client
  */
@@ -55,7 +57,7 @@ public abstract class DefaultClientConfig implements IClientConfig {
     /**
      * The serializer instance
      */
-    private DefaultSerializer serializer;
+    private ISerializer serializer;
 
     /**
      * Creates an instance of this configuration with an authentication provider
@@ -97,11 +99,22 @@ public abstract class DefaultClientConfig implements IClientConfig {
      */
     @Override
     public IHttpProvider getHttpProvider() {
+        return this.getHttpProvider(null);
+    }
+
+    /**
+     * Gets the HTTP provider
+     *
+     * @param httpClient the http client to pass to the http provider when building it
+     * @param <T1> the http client type
+     * @return the HTTP provider
+     */
+    @Override
+    public <T1> IHttpProvider getHttpProvider(final T1 httpClient) {
         if (httpProvider == null) {
-            httpProvider = new CoreHttpProvider(getSerializer(),
-                    getAuthenticationProvider(),
-                    getExecutors(),
-                    getLogger());
+            httpProvider = (httpClient instanceof OkHttpClient) ? 
+                            new CoreHttpProvider(this, (OkHttpClient) httpClient) :
+                            new CoreHttpProvider(this, null);
             getLogger().logDebug("Created CoreHttpProvider");
         }
         return httpProvider;

--- a/src/main/java/com/microsoft/graph/core/IClientConfig.java
+++ b/src/main/java/com/microsoft/graph/core/IClientConfig.java
@@ -54,6 +54,15 @@ public interface IClientConfig {
     IHttpProvider getHttpProvider();
 
     /**
+     * Gets the HTTP provider
+     * 
+     * @param httpClient the http client to pass to the http provider when building it
+     * @param <T1> the http client type
+     * @return the HTTP provider
+     */
+    <T1> IHttpProvider getHttpProvider(final T1 httpClient);
+
+    /**
      * Gets the logger
      * 
      * @return the logger

--- a/src/main/java/com/microsoft/graph/http/CoreHttpProvider.java
+++ b/src/main/java/com/microsoft/graph/http/CoreHttpProvider.java
@@ -46,6 +46,7 @@ import com.microsoft.graph.concurrency.IProgressCallback;
 import com.microsoft.graph.core.ClientException;
 import com.microsoft.graph.core.Constants;
 import com.microsoft.graph.core.DefaultConnectionConfig;
+import com.microsoft.graph.core.IClientConfig;
 import com.microsoft.graph.core.IConnectionConfig;
 import com.microsoft.graph.httpcore.HttpClients;
 import com.microsoft.graph.httpcore.ICoreAuthenticationProvider;
@@ -120,6 +121,18 @@ public class CoreHttpProvider implements IHttpProvider {
 	}
 
 	/**
+	 * Creates the DefaultHttpProvider
+	 *
+	 * @param clientConfig           the client configuration to use for the provider
+	 * @param httpClient             the http client to execute the requests with
+	 */
+	public CoreHttpProvider(final IClientConfig clientConfig,
+			final OkHttpClient httpClient) {
+		this(clientConfig.getSerializer(), clientConfig.getAuthenticationProvider(), clientConfig.getExecutors(), clientConfig.getLogger());
+		this.corehttpClient = httpClient;
+	}
+
+	/**
 	 * Gets the serializer for this HTTP provider
 	 *
 	 * @return the serializer for this provider
@@ -144,9 +157,9 @@ public class CoreHttpProvider implements IHttpProvider {
 			final ICallback<? super Result> callback,
 			final Class<Result> resultClass,
 			final Body serializable) {
-		final IProgressCallback<Result> progressCallback;
+		final IProgressCallback<? super Result> progressCallback;
 		if (callback instanceof IProgressCallback) {
-			progressCallback = (IProgressCallback<Result>) callback;
+			progressCallback = (IProgressCallback<? super Result>) callback;
 		} else {
 			progressCallback = null;
 		}
@@ -221,7 +234,7 @@ public class CoreHttpProvider implements IHttpProvider {
 	public <Result, Body> Request getHttpRequest(final IHttpRequest request,
 			final Class<Result> resultClass,
 			final Body serializable,
-			final IProgressCallback<Result> progress) throws ClientException {
+			final IProgressCallback<? super Result> progress) throws ClientException {
 		final int defaultBufferSize = 4096;
 
 		final URL requestUrl = request.getRequestUrl();
@@ -354,7 +367,7 @@ public class CoreHttpProvider implements IHttpProvider {
 	private <Result, Body, DeserializeType> Result sendRequestInternal(final IHttpRequest request,
 			final Class<Result> resultClass,
 			final Body serializable,
-			final IProgressCallback<Result> progress,
+			final IProgressCallback<? super Result> progress,
 			final IStatefulResponseHandler<Result, DeserializeType> handler)
 					throws ClientException {
 
@@ -374,11 +387,11 @@ public class CoreHttpProvider implements IHttpProvider {
 									.newBuilder()
 									.connectTimeout(connectionConfig.getConnectTimeout(), TimeUnit.MILLISECONDS)
 									.readTimeout(connectionConfig.getReadTimeout(), TimeUnit.MILLISECONDS)
-									.followRedirects(false)
+									.followRedirects(false) //TODO those are duplicated for the time being with core, remove in 3.0
 									.protocols(Arrays.asList(Protocol.HTTP_1_1)) //https://stackoverflow.com/questions/62031298/sockettimeout-on-java-11-but-not-on-java-8
 									.build();
 			}
-			if (authenticationProvider != null) {
+			if (authenticationProvider != null) { //TODO remove the authentication provider as this is done by the interceptor anyway
 				authenticationProvider.authenticateRequest(request);
 			}
 			Request coreHttpRequest = getHttpRequest(request, resultClass, serializable, progress);

--- a/src/main/java/com/microsoft/graph/http/CoreHttpProvider.java
+++ b/src/main/java/com/microsoft/graph/http/CoreHttpProvider.java
@@ -387,11 +387,11 @@ public class CoreHttpProvider implements IHttpProvider {
 									.newBuilder()
 									.connectTimeout(connectionConfig.getConnectTimeout(), TimeUnit.MILLISECONDS)
 									.readTimeout(connectionConfig.getReadTimeout(), TimeUnit.MILLISECONDS)
-									.followRedirects(false) //TODO those are duplicated for the time being with core, remove in 3.0
+									.followRedirects(false) //TODO https://github.com/microsoftgraph/msgraph-sdk-java/issues/516
 									.protocols(Arrays.asList(Protocol.HTTP_1_1)) //https://stackoverflow.com/questions/62031298/sockettimeout-on-java-11-but-not-on-java-8
 									.build();
 			}
-			if (authenticationProvider != null) { //TODO remove the authentication provider as this is done by the interceptor anyway
+			if (authenticationProvider != null) { //TODO https://github.com/microsoftgraph/msgraph-sdk-java/issues/517
 				authenticationProvider.authenticateRequest(request);
 			}
 			Request coreHttpRequest = getHttpRequest(request, resultClass, serializable, progress);

--- a/src/main/java/com/microsoft/graph/http/DefaultHttpProvider.java
+++ b/src/main/java/com/microsoft/graph/http/DefaultHttpProvider.java
@@ -51,7 +51,9 @@ import java.util.Scanner;
 
 /**
  * HTTP provider based off of URLConnection
+ * @deprecated user CoreHttpProvider instead
  */
+@Deprecated
 public class DefaultHttpProvider implements IHttpProvider {
 
     /**
@@ -160,9 +162,9 @@ public class DefaultHttpProvider implements IHttpProvider {
                                     final ICallback<? super Result> callback,
                                     final Class<Result> resultClass,
                                     final Body serializable) {
-        final IProgressCallback<Result> progressCallback;
+        final IProgressCallback<? super Result> progressCallback;
         if (callback instanceof IProgressCallback) {
-            progressCallback = (IProgressCallback<Result>) callback;
+            progressCallback = (IProgressCallback<? super Result>) callback;
         } else {
             progressCallback = null;
         }
@@ -241,7 +243,7 @@ public class DefaultHttpProvider implements IHttpProvider {
     private <Result, Body, DeserializeType> Result sendRequestInternal(final IHttpRequest request,
                                                                        final Class<Result> resultClass,
                                                                        final Body serializable,
-                                                                       final IProgressCallback<Result> progress,
+                                                                       final IProgressCallback<? super Result> progress,
                                                                        final IStatefulResponseHandler<Result, DeserializeType> handler)
             throws ClientException {
         final int defaultBufferSize = 4096;
@@ -535,7 +537,7 @@ public class DefaultHttpProvider implements IHttpProvider {
 
 	@Override
 	public <Result, BodyType> Request getHttpRequest(IHttpRequest request, Class<Result> resultClass,
-			BodyType serializable, IProgressCallback<Result> progress) throws ClientException {
+			BodyType serializable, IProgressCallback<? super Result> progress) throws ClientException {
 		return null;
 	}
 }

--- a/src/main/java/com/microsoft/graph/http/IHttpProvider.java
+++ b/src/main/java/com/microsoft/graph/http/IHttpProvider.java
@@ -123,6 +123,6 @@ public interface IHttpProvider {
     <Result, BodyType> Request getHttpRequest(final IHttpRequest request,
                                               final Class<Result> resultClass,
                                               final BodyType serializable,
-                                              final IProgressCallback<Result> progress)
+                                              final IProgressCallback<? super Result> progress)
             throws ClientException;
 }

--- a/src/test/java/com/microsoft/graph/http/MockHttpProvider.java
+++ b/src/test/java/com/microsoft/graph/http/MockHttpProvider.java
@@ -129,7 +129,7 @@ public class MockHttpProvider implements IHttpProvider {
 
 	@Override
 	public <Result, BodyType> Request getHttpRequest(IHttpRequest request, Class<Result> resultClass,
-			BodyType serializable, IProgressCallback<Result> progress) throws ClientException {
+			BodyType serializable, IProgressCallback<? super Result> progress) throws ClientException {
 		return null;
 	}
 }

--- a/src/test/java/com/microsoft/graph/requests/extensions/GraphServiceClientTest.java
+++ b/src/test/java/com/microsoft/graph/requests/extensions/GraphServiceClientTest.java
@@ -207,7 +207,7 @@ public class GraphServiceClientTest {
 
 			@Override
 			public <Result, BodyType> Request getHttpRequest(IHttpRequest request, Class<Result> resultClass,
-					BodyType serializable, IProgressCallback<Result> progress) throws ClientException {
+					BodyType serializable, IProgressCallback<? super Result> progress) throws ClientException {
 				return null;
 			}
         };

